flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/maintenance/stream/ExpireSnapshots.java

```diff
@@ -0,0 +1,161 @@
+
+  /** Creates the builder for creating a stream which expires snapshots for the table. */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  public static class Builder extends MaintenanceTaskBuilder<ExpireSnapshots.Builder> {
+    private Duration minAge = null;
+    private Integer retainLast = null;
+    private int planningWorkerPoolSize = DELETE_PLANNING_WORKER_POOL_SIZE_DEFAULT;
+    private int deleteAttemptNum = DELETE_ATTEMPT_NUM;
+    private int deleteWorkerPoolSize = DELETE_WORKER_POOL_SIZE_DEFAULT;
+
+    /**
+     * The snapshots newer than this age will not be removed.
+     *
+     * @param newMinAge of the files to be removed
+     * @return for chained calls
+     */
+    public Builder minAge(Duration newMinAge) {
+      this.minAge = newMinAge;
+      return this;
+    }
+
+    /**
+     * The minimum {@link org.apache.iceberg.Snapshot}s to retain. For more details description see
+     * {@link org.apache.iceberg.ExpireSnapshots#retainLast(int)}.
+     *
+     * @param newRetainLast number of snapshots to retain
+     * @return for chained calls
+     */
+    public Builder retainLast(int newRetainLast) {
+      this.retainLast = newRetainLast;
+      return this;
+    }
+
+    /**
+     * The worker pool size used to calculate the files to delete.
+     *
+     * @param newPlanningWorkerPoolSize for planning files to delete
+     * @return for chained calls
+     */
+    public Builder planningWorkerPoolSize(int newPlanningWorkerPoolSize) {
+      this.planningWorkerPoolSize = newPlanningWorkerPoolSize;
+      return this;
+    }
+
+    /**
+     * The number of retries on the failed delete attempts.
+     *
+     * @param newDeleteAttemptNum number of retries
+     * @return for chained calls
+     */
+    public Builder deleteAttemptNum(int newDeleteAttemptNum) {
+      this.deleteAttemptNum = newDeleteAttemptNum;
+      return this;
+    }
+
+    /**
+     * The worker pool size used for deleting files.
+     *
+     * @param newDeleteWorkerPoolSize for scanning
+     * @return for chained calls
+     */
+    public Builder deleteWorkerPoolSize(int newDeleteWorkerPoolSize) {
+      this.deleteWorkerPoolSize = newDeleteWorkerPoolSize;
+      return this;
+    }
+
+    @Override
+    DataStream<TaskResult> buildInternal(DataStream<Trigger> trigger) {
+      Preconditions.checkNotNull(tableLoader(), "TableLoader should not be null");
+
+      SingleOutputStreamOperator<TaskResult> result =
+          trigger
+              .process(
+                  new ExpireSnapshotsProcessor(
+                      tableLoader(),
+                      minAge == null ? null : minAge.toMillis(),
+                      retainLast,
+                      planningWorkerPoolSize))
+              .name(EXECUTOR_TASK_NAME)
+              .uid(uidPrefix() + "-expire-snapshots")
+              .slotSharingGroup(slotSharingGroup())
+              .forceNonParallel();
+
+      AsyncRetryStrategy<Boolean> retryStrategy =
```

Comment:
do we still need async with the batch/bulk delete API (as suggested in another comment)?

Reply:
See above